src/nn/embedding.rs

```diff
@@ -0,0 +1,230 @@
+use crate::{gradients::Tape, optim::*, shapes::*, tensor::*, tensor_ops::*};
+
+use super::module::{Module, ModuleMut, ResetParams};
+
+/// An embedding
+/// Initializes [Self::weight] from a Uniform distribution
+/// between [-1 / sqrt(I), 1 / sqrt(I)].
+///
+/// # Generics
+/// - `VOCAB` The size of the vocabulary, inputs integer values must be between
+///    0 and VOCAB;
+/// - `DIM` The "output" size of vectors & matrices which are the vectors being selected.
+///
+/// # Examples
+/// `Embedding<5, 2>` can act on vectors with SEQ integer elements (with values between 0 and 4), and results in a SEQ tensor of
+/// usually f32 elements being the rows in [Self::weight].
+/// ```rust
+///
+/// # use dfdx::prelude::*;
+/// # let dev: Cpu = Default::default();
+/// let model: Embedding<7, 2> = dev.build_module();
+/// // single sequence of ids
+/// let inputs: Tensor<_, usize> = dev.zeros::<Rank1<5>>();
+/// let _: Tensor<(Const<5>, Const<2>,), f32> = model.forward(inputs);
+/// // batched sequence of ids
+/// let inputs: Tensor<_, usize> = dev.zeros::<Rank2<10, 5>>();
+/// let _: Tensor<(Const<10>, Const<5>, Const<2>), f32> = model.forward(inputs);
+/// ```
+#[derive(Debug, Clone)]
+pub struct Embedding<const VOCAB: usize, const DIM: usize, D: Device<f32> = Cpu> {
+    /// Transposed weight matrix, shape (I, O)
+    pub weight: Tensor<Rank2<VOCAB, DIM>, f32, D>,
+}
+
+impl<const VOCAB: usize, const DIM: usize, const SEQ: usize, D: Device<f32>, T: Tape<D>>
+    Module<Tensor<Rank1<SEQ>, usize, D, T>> for Embedding<VOCAB, DIM, D>
+{
+    type Output = Tensor<Rank2<SEQ, DIM>, f32, D, T>;
+    fn forward(&self, input: Tensor<Rank1<SEQ>, usize, D, T>) -> Self::Output {
+        self.weight.retaped::<T>().gather(input)
+    }
+}
+
+impl<
+        const VOCAB: usize,
+        const DIM: usize,
+        const SEQ: usize,
+        const BATCH: usize,
+        D: Device<f32>,
+        T: Tape<D>,
+    > Module<Tensor<Rank2<BATCH, SEQ>, usize, D, T>> for Embedding<VOCAB, DIM, D>
+{
+    type Output = Tensor<Rank3<BATCH, SEQ, DIM>, f32, D, T>;
+    fn forward(&self, input: Tensor<Rank2<BATCH, SEQ>, usize, D, T>) -> Self::Output {
+        self.weight.retaped::<T>().gather(input)
```

Comment:
Also, do you still need the retaped even with the modifications to gather?

Reply:
Not with the manual split_tape.